pkg/sql/evalcatalog/pg_updatable.go

```diff
@@ -33,6 +38,43 @@ var (
 	allUpdatableEvents = tree.NewDInt((1 << UpdateCommand) | (1 << InsertCommand) | (1 << DeleteCommand))
 )
 
+// RedactDescriptor takes an encoded protobuf descriptor and returns the
+// encoded protobuf after redaction.
+func (b *Builtins) RedactDescriptor(ctx context.Context, encodedDescriptor []byte) ([]byte, error) {
+	var desc descpb.Descriptor
+	if err := protoutil.Unmarshal(encodedDescriptor, &desc); err != nil {
+		return nil, err
+	}
+	// The descpb package has some strict rules about the timestamps in the
+	// descriptors. The strictness was an attempt to make it difficult to read the
+	// descriptor without setting the MVCC timestamp from the descriptor. In order
+	// to make it safe to use the descriptor, we preempt the assertions by setting
+	// a non-empty timestamp if there is no timestamp set. This is always safe.
+	//
+	// A downside of setting this timestamp is that it'll be set in the returned
+	// descriptor. There is no existing uniform mechanism to reset the
+	// ModificationTime across the protobuf descriptor oneof members, and this
+	// seems like it's probably a good thing. At some point we may want to
+	// consider replacing all of this policy with something more flexible. For
+	// now, the silly sentinel timestamp should serve to make these descriptors
+	// harder to use, which seems good.
```

Comment:
I think we should go ahead and fix the logic in `descpb`. Such as it is written, it predates the wrapping of protobufs behind interfaces, the builder, the unification of the storage read paths, etc.

At the very least, the `descpb` logic should not panic if it's not happy about the timestamp. I think I would prefer, however, replacing `descpb.FromDescriptor*` with something that takes a `kv.KeyValue` that's expected to contain a valid serialized protobuf + timestamp and returns an error as an extra 5th return value. I'll see what I can do.

Reply:
Follow up: https://github.com/cockroachdb/cockroach/pull/88786
It's probably worth rebasing this PR on mine.